File: 류호석배 알고리즘 코딩 테스트/제1회/4번-꿈틀꿈틀 호석 애벌레/subsolution(brute force).cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
typedef long long int ll;
#define INF 0x7fffffff
#define NM 100005

using namespace std;

int n, K, D;
ll a[NM], ans;
void input() {
	cin >> n >> K;
	D = 0;
	for (int i = 1; i <= n; i++) cin >> a[i];
}

void dfs(int x, ll energy) {
	if (x > n) {
		ans = max(ans, energy);
		return;
	}
	ll sum = 0;

	// x 번 먹이를 먹기로 결심!
	for (int i = x; i <= n; i++) {
		sum += a[i];
		if (sum >= K) {
			dfs(i + D + 1, energy + sum - K);
			break;
		}
	}

	// x 번 먹이를 안 먹기로 결심!
	dfs(x + 1, energy);
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	input();
	dfs(1, 0);
	cout << ans;
	return 0;
}

File: 류호석배 알고리즘 코딩 테스트/제1회/4번-꿈틀꿈틀 호석 애벌레/subsolution(brute force).java
import java.io.*;
import java.util.*;


public class Main {
    static int n, K, D;
    static long[] a, dy;
    static long ans;
    static FastReader scan = new FastReader();
    static PrintWriter out = new PrintWriter(System.out);

    static void input() {
        n = scan.nextInt();
        K = scan.nextInt();
        a = new long[n + 1];
        dy = new long[n + 1];
        for (int i = 1; i <= n; i++) {
            a[i] = scan.nextLong();
        }
    }

    static void dfs(int idx, long energy) {
        ans = Math.max(ans, energy);

        long sum = 0;
        for (int r = idx; r <= n; r++) {
            sum += a[r];
            if (sum >= K) {
                dfs(r + 1, energy + (sum - K));
                break;
            }
        }
        dfs(idx + 1, energy);
    }

    static void pro() {
        dfs(1, 0);
        out.print(ans);
        out.close();
    }

    public static void main(String[] args) {
        input();
        pro();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}